pkgs/circuits/halo2-secp256k1-wasm/src/eff_ecdsa.rs

```diff
@@ -0,0 +1,402 @@
+        let ecdsa_inputs = EffECDSAInputs {
+            s: ecdsa_inputs.s,
+            pk: ecdsa_inputs.pk,
+            T: ecdsa_inputs.T,
+            U: ecdsa_inputs.U,
+        };
+
+        let circuit_params = halo2_wasm
+            .circuit_params
+            .clone()
+            .context("Error: Circuit params are not set")?;
+
+        let lookup_bits = circuit_params
+            .lookup_bits
+            .context("Error: Lookup bits are not set in circuit params")?;
+
+        let range = RangeChip::new(
+            lookup_bits,
+            halo2_wasm.circuit.borrow().lookup_manager().clone(),
+        );
+
+        Ok(Secp256k1VerifyCircuit {
+            eff_ecdsa_inputs: ecdsa_inputs,
+            range_chip: range,
+            builder: Rc::clone(&halo2_wasm.circuit),
+        })
+    }
+
+    fn secp256k1_fp_chip(&self) -> Secp256k1FpChip<Bn254Fr> {
+        let limb_bits = 88;
+        let num_limbs = 3;
+        Secp256k1FpChip::<Bn254Fr>::new(&self.range_chip, limb_bits, num_limbs)
+    }
+    fn secp256k1_fq_chip(&self) -> Secp256k1FqChip<Bn254Fr> {
+        let limb_bits = 88;
+        let num_limbs = 3;
+        Secp256k1FqChip::<Bn254Fr>::new(&self.range_chip, limb_bits, num_limbs)
+    }
+
+    fn ecc_chip<'a, Fp: BigPrimeField>(
+        &'a self,
+        fp_chip: &'a FpChip<Bn254Fr, Secp256k1Fp>, // TODO that could be generalized in the future
+    ) -> EccChip<'a, Bn254Fr, FpChip<Bn254Fr, Secp256k1Fp>> {
+        EccChip::new(fp_chip)
+    }
+
+    fn recover_pk_eff<F, CF, SF, GA>(
+        &self,
+        ecc_chip: &EccChip<F, FpChip<F, CF>>,
+        ctx: &mut Halo2Context<F>,
+        T: EcPoint<F, <FpChip<F, CF> as FieldChip<F>>::FieldPoint>,
+        U: EcPoint<F, <FpChip<F, CF> as FieldChip<F>>::FieldPoint>,
+        s: ProperCrtUint<F>,
+        fixed_window_bits: usize,
+    ) -> EcPoint<F, <FpChip<F, CF> as FieldChip<F>>::FieldPoint>
+    where
+        F: BigPrimeField,
+        CF: BigPrimeField,
+        SF: BigPrimeField,
+        GA: CurveAffineExt<Base = CF, ScalarExt = SF>,
+    {
+        // Following https://en.wikipedia.org/wiki/Elliptic_Curve_Digital_Signature_Algorithm
+        let base_chip = ecc_chip.field_chip;
+        let scalar_chip =
+            FpChip::<F, SF>::new(base_chip.range, base_chip.limb_bits, base_chip.num_limbs);
+
+        // Check s is in [1, (n-1)]
+        scalar_chip.is_soft_nonzero(ctx, &s);
+        base_chip.enforce_less_than(ctx, T.x().clone());
+        base_chip.enforce_less_than(ctx, U.x().clone());
+
+        // Recover the public key from signature
+        // s_mul_t = s * T
+        let s_mul_t = ecc_chip.scalar_mult::<GA>(
+            ctx,
+            T,
+            s.limbs().to_vec(),
+            base_chip.limb_bits,
+            fixed_window_bits,
+        );
+
+        // s_mul_t + U = pk
+        let recovered_pk = ecc_chip.add_unequal(ctx, &s_mul_t, &U, false);
+
+        recovered_pk
```

Comment:
unnecessary let binding

```suggestion
        ecc_chip.add_unequal(ctx, &s_mul_t, &U, false)
```

Reply:
That is fixed in another PR.